openedx/core/djangoapps/agreements/api.py

```diff
@@ -66,5 +78,12 @@ def get_integrity_signatures_for_course(course_id):
     Returns:
         * QuerySet of IntegritySignature objects (can be empty).
     """
+
     course_key = CourseKey.from_string(course_id)
-    return IntegritySignature.objects.filter(course_key=course_key)
+    course_integrity_signature = IntegritySignature.objects.filter(
+        course_key=course_key
+    ).select_related('user')
+    for signature in course_integrity_signature:
+        cache_key = get_integrity_signature_cache_key(signature.user.username, course_id)
+        cache.set(cache_key, signature)
+    return course_integrity_signature
```

Comment:
The caching here is over-aggressive, and will cause trouble for larger courses. It's forcing a complete iteration of the queryset before the caller of this function can even start iterating through the results. Worse, each step of the iteration is going to be blocked by a synchronous request to set the cached value in memcached, taking somewhere around 1 ms per item.

Say this queryset had 10K records in it. The row sizes are very small, so that can execute on MySQL and come across the wire in maybe 10 ms. Prior to this new caching, that's when the caller of this function would start being able to iterate through it. But add in the caching, and now we've added 10,000 ms of time setting cache values in memcached, one entry at a time, and made this function return 1000X slower.